File: Question6JS/question6test.js
const { Builder, By, until } = require('selenium-webdriver');
const assert = require('assert');

async function runTest() {
  	const base_url = "https://parabank.parasoft.com/parabank/index.htm";
  	
  	// should be parameterized from command line
  	const username = process.env.TEST_USERNAME || 'john';
  	const password = process.env.TEST_PASSWORD || 'demo';
  	
  	// parameter to run tests headless
	const headless = process.env.TEST_HEADLESS !== 'false'; // default is true  	
  	
  	// number of attempts to refresh accounts table
	let attemptCount = 3;


  	// browser options
  	const chrome = require('selenium-webdriver/chrome');
  	let options = new chrome.Options();
	options.addArguments("--window-size=1920,1080");
	if (headless)
  		options.addArguments("--headless");

  	// driver
  	const driver = await new Builder()
    	.forBrowser('chrome')
    	.setChromeOptions(options)
    	.build();

  	try {
    	
    	// go to the site
    	await driver.get(base_url);
    	const title = await driver.getTitle();
    	assert(title.includes("ParaBank"), "Page title is incorrect");

    	// login
    	await driver.findElement(By.name('username')).sendKeys(username);
    	await driver.findElement(By.name('password')).sendKeys(password);
    	await driver.findElement(By.css('input.button')).click();

    	// wait for account overview table
    	await driver.wait(until.elementLocated(By.id('accountTable')), 10000);
		const overview = await driver.findElement(By.id('showOverview')).getText();
		assert(overview.includes("Accounts Overview"), 
			"Expected 'Accounts Overview' to appear, but got: " + overview);
    	
    	// getting accounts links
    	let accountLinks = await driver.findElements(By.xpath("//td/a"));

    	// click "Accounts Overview" again if there's only one account
    	let attempts = 0;
    	while (accountLinks.length <= 1 && attempts < attemptCount) {
    		console.log("Refreshing the account list...");
      		await driver.findElement(By.linkText('Accounts Overview')).click();
      		await driver.sleep(5000);
      		accountLinks = await driver.findElements(By.xpath("//td/a"));
      		attempts++;
    	}

    	// Loop through accounts
    	for (let i = 0; i < accountLinks.length; i++) {
      	try {
        	const accountId = await accountLinks[i].getText();
        	await accountLinks[i].click();

			// waiting for account id on a page
      		await driver.sleep(5000);
        	let accountDetails = await driver.wait(
          		until.elementLocated(By.id('accountId')), 5000);
        	const textDetails = await driver.findElement(By.id('accountDetails')).getText();
			assert(textDetails.includes(accountId), 
				"Expected " + accountId + " to appear, but got: " + textDetails);
 
        	// getting the last transaction element from the table
        	const transactionRow = await driver.findElement(
          		By.xpath("//table[@id='transactionTable']/tbody/tr[last()]")
        	);
        	const lastTransaction = await transactionRow.getText();
        	console.log('Last transaction found:', lastTransaction);

        	// verify row starts with a valid date
        	assert(/^\d{2}-\d{2}-\d{4}.*/.test(lastTransaction), 
        		"Row does not start with a valid date");

        	// verify there is only one currency value in the string
        	const countValues = (lastTransaction.match(/\$\d+\.\d{2}/g) || []).length;
			assert.strictEqual(countValues, 1, "Unexpected number of currency values");
        	
        	// check the description is not empty
        	const descriptionStart = lastTransaction.indexOf(" ")+ 1;
        	const amountStart = lastTransaction.lastIndexOf('$');
        	const description = lastTransaction.substring(descriptionStart, amountStart).trim();
			assert(description.length > 0, "Description is empty");

        	// exit from loop
        	break;

      	} catch (error) {
        	console.log('Skipping glitchy account');
        	await driver.navigate().back();
        	await driver.wait(until.elementLocated(By.xpath("//td/a")), 5000);
        	accountLinks = await driver.findElements(By.xpath("//td/a")); // re-fetch
      	}
    }

  } catch (error) {
    	console.log("No transactions found. "
					+ "This may be expected for a new or inactive account.");
  } finally {
    await driver.quit();
  }
}

runTest();

File: AssessmentProject/src/test/java/com/ssc/Assessment/tests/Question6Test.java
package com.ssc.Assessment.tests;

import static org.junit.Assert.*;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.chrome.ChromeDriver;
import org.openqa.selenium.chrome.ChromeOptions;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;

import net.serenitybdd.junit.runners.SerenityRunner;

import java.time.Duration;
import java.util.List;


@RunWith(SerenityRunner.class)
public class Question6Test {
	
	private static final String BASE_URL = "https://parabank.parasoft.com/parabank/index.htm";

	@Test
	public void verifyLatestTransactionDisplayedCorrectly() {
		
		// should be parameterized from command line
		String username = System.getProperty("test.username", "john");
		String password = System.getProperty("test.password", "demo");
		
		// parameter to run tests headless
		String headless = System.getProperty("test.headless", "true");

		// number of attempts to refresh accounts table
		int attemptCount = 3;

		
		// browser options
		ChromeOptions options = new ChromeOptions();
		options.addArguments("--window-size=1920,1080");
		if (headless.equalsIgnoreCase("true")) {
		    options.addArguments("--headless");
		}
		
		// webdriver
		WebDriver driver = new ChromeDriver(options);
		WebDriverWait wait = new WebDriverWait(driver, Duration.ofSeconds(10));
        List<WebElement> accountLinks = null; // list of account links
		
		try {

			// go to the site
            driver.get(BASE_URL);
            assertTrue(driver.getTitle().contains("ParaBank"));
            
            // login
            driver.findElement(By.name("username")).sendKeys(username);
            driver.findElement(By.name("password")).sendKeys(password);
            driver.findElement(By.cssSelector("input.button")).click();
            
 
            // wait for account overview Table
            wait.until(ExpectedConditions.visibilityOfElementLocated(By.id("accountTable")));
            assertTrue(driver.findElement(By.id("showOverview"))
            		.getText().contains("Accounts Overview"));
            

            // because not all accounts visible, click on link "Accounts Overview"
            // and wait till there is more accounts than 1
            for (int i = 0; i < attemptCount ; i++) {
            	try {
            		accountLinks = wait.until(d -> {
            			List<WebElement> links = d.findElements(By.xpath("//td/a"));
            			return (links.size() > 1) ? links : null;
            		});
            	} catch (Exception e) {
            		System.out.println("Not enough accounts are visible");
            	}
            
            	// if still only 1 link, try clicking again
            	if ((accountLinks == null || accountLinks.size() <=1) && i <attemptCount - 1) {
            		driver.findElement(By.linkText("Accounts Overview")).click();
            		System.out.println("Refreshing the list...");
            	}
            }
      
            
            // looping through accounts, some of them might not work
            accountLinks = driver.findElements(By.xpath("//td/a"));
            for (WebElement accountLink : accountLinks) {
            	
            	try {
                    String accountId = accountLink.getText();
                    accountLink.click();
                    
                    // waiting for account id on a page
                    wait.until(d -> {
                        WebElement el = d.findElement(By.id("accountId"));
                        return !el.getText().trim().isEmpty();
                    });
                    assertTrue(driver.findElement(By.id("accountDetails"))
                    		.getText().contains(accountId));

                    // getting the last transaction element from the table
                    WebElement lastTransactionEl = driver.findElement
                    		(By.xpath("//table[@id='transactionTable']/tbody/tr[last()]"));
                    String lastTransaction = lastTransactionEl.getText();
                    System.out.println("Last Transaction: " + lastTransaction);
                    
                    
                    // verify row starts with a valid date
                    assertTrue("Row does not start with a valid date", lastTransaction.matches("^\\d{2}-\\d{2}-\\d{4}.*"));
 
                    // verify there is only one currency value in the string
                    int countValues = lastTransaction.split("\\$\\d+\\.\\d{2}", -1).length - 1;
                    assertTrue("Unexpected number of Currency Values", countValues==1);
                    
                    // check the description is not empty
                    int descriptionStart = lastTransaction.indexOf(" ") + 1;
                    int amountStart = lastTransaction.lastIndexOf("$");
                    assertFalse("Description is empty", 
                    		lastTransaction.substring(descriptionStart, amountStart).trim().isEmpty());
                    
                    // exit from loop
                    break;
                   		
            	} catch (Exception e) {
            		System.out.println("Skipping glitchy account...");
                    driver.navigate().back();
                    wait.until(ExpectedConditions.numberOfElementsToBeMoreThan(
                    	    By.xpath("//td/a"), 0));
                    accountLinks = driver.findElements(By.xpath("//td/a"));              		
            	}
            	
            }

		} catch (Exception e) {
 
			System.out.println("No transactions found. "
					+ "This may be expected for a new or inactive account.");
            
		} finally {
			
			driver.quit();
		}
		
	}

}
